projects/sandbox/datagen/datagen/utils/timeslides.py

```diff
@@ -70,23 +78,62 @@ def __call__(self, parameters):
         )
 
 
-def _generate_waveforms(sampler, generator):
-    params = sampler()
-    waveforms = generator(params)
-    return waveforms, params
+def _generate_waveforms(
+    sampler: Sampler,
+    generator: WaveformGenerator,
+    hopeless_snr_threshold: float,
+    psds: "torch.Tensor",
+    sample_rate: float,
+    highpass: float,
+):
+    waveforms = []
+    n_rejected = 0
+    while len(waveforms) < sampler.num_signals:
+        params = sampler()
+        signals = generator(params)
+
+        # crude estimate of snrs using hplus
+        snrs = compute_network_snr(
+            signals[:, 0.0:], psds, sample_rate, highpass
+        )
+        signals = signals[snrs > hopeless_snr_threshold]
+        n_rejected += np.sum([snrs < hopeless_snr_threshold])
+        waveforms.append(signals)
```

Comment:
We could end up with a lot more signals than can fit in the segment with the desired spacing. We'll also need a way to make sure that signals don't overlap.

Reply:
All of the logic that determines the number of signals based on the segment length and spacing is already controlled by the `Sampler`

Reply:
I'm missing something - how does that work? If you reject, say, 25% of the signals generated each time, won't `len(waveforms)` be `1.5 * num_signals`?

Reply:
I see the confusion. This commit doesnt have the `waveforms = waveforms[:n_signals]` thats on my local repo.

Reply:
Ah okay, I see